### ISSUE TITLE
Memory leak in terminal

### ISSUE THREAD

**Issue by s1n7ax** [neovim/neovim#15651]
### Neovim version (nvim -v)

NVIM v0.5.0 Build type: Release LuaJIT 2.0.5

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Arch Linux: up to date: 13-09-21

### Terminal name/version

suckless st

### $TERM environment variable

st-256color

### Installation

pacman official repositories

### How to reproduce the issue

* Add following keymap

```
 {'t', '<c-n>', [[<c-\><c-n>]], {}},
```

* Open neovim terminal
* Execute the key combination in terminal insert mode

### More information:

```
:tmap <c-n>
t  <C-N>         <C-\><C-N>
```

Just executing ` <C-\><C-N>` will result in same behavior, how ever, If I remap the same action to any other key combination if works fine.

https://user-images.githubusercontent.com/18459807/133034264-c94a4a0c-9362-4445-88a5-7ce6a3a15f75.mp4

 

### Expected behavior

The mode should be switched to normal mode

### Actual behavior

Neovim freezes and will not react to any key presses after. The CPU and memory usage will increase until no memory left and freezes the OS.

**Comment by zeertzjq** [comment]
Vim seems to also behave the same. What's more, this applies any any recursive mapping in terminal mode without `noremap`.

**Comment by zeertzjq** [comment]
Maybe this is the expected behavior? There is a similar closed issue: #13810 (https://github.com/vim/vim/issues/7716).

**Comment by gpanders** [comment]
Yes this does not appear to be a bug, but simply a recursive map. Maybe we should catch these after a certain recursion depth rather than freezing indefinitely though.

**Comment by justinmk** [comment]
there is the 'maxmapdepth' option but as usual it's a half-measure:

> 'maxmapdepth' 'mmd'	number	(default 1000)
> 			global
> 	Maximum number of times a mapping is done without resulting in a
> 	character to be used.  This normally catches endless mappings, like
> 	":map x y" with ":map y x".  **It still does not catch ":map g wg",
> 	because the 'w' is used before the next mapping is done.**

@s1n7ax does `:set maxmapdepth=10` avoid the freeze?

**Comment by s1n7ax** [comment]
> does :set maxmapdepth=10 avoid the freeze?

Nope
